contracts/puppeteer-initia/src/contract.rs

```diff
@@ -0,0 +1,1423 @@
+    let config = puppeteer_base.config.load(deps.storage)?;
+    validate_sender(&config, &info.sender)?;
+    puppeteer_base.validate_tx_idle_state(deps.as_ref())?;
+    // exclude fees, no need to send local denom tokens to remote zone
+    let message_funds: Vec<_> = info
+        .funds
+        .into_iter()
+        .filter(|f| f.denom != LOCAL_DENOM)
+        .collect();
+    ensure_eq!(
+        message_funds.len(),
+        1,
+        ContractError::InvalidFunds {
+            reason: "Only one coin is allowed".to_string()
+        }
+    );
+    let coin = message_funds.first().ok_or(ContractError::InvalidFunds {
+        reason: "No funds".to_string(),
+    })?;
+    let ica_address = puppeteer_base.ica.get_address(deps.storage)?;
+    let msg = NeutronMsg::IbcTransfer {
+        source_port: config.port_id,
+        source_channel: config.transfer_channel_id,
+        token: (*coin).clone(),
+        sender: env.contract.address.to_string(),
+        receiver: ica_address.to_string(),
+        timeout_height: RequestPacketTimeoutHeight {
+            revision_number: None,
+            revision_height: None,
+        },
+        timeout_timestamp: env.block.time.plus_seconds(config.timeout).nanos(),
+        memo: "".to_string(),
+        fee: query_ibc_fee(deps.as_ref(), LOCAL_DENOM)?,
+    };
+    let submsg = puppeteer_base.msg_with_sudo_callback(
+        deps,
+        msg,
+        Transaction::IBCTransfer {
+            denom: coin.denom.to_string(),
+            amount: coin.amount.into(),
+            reason,
+            recipient: ica_address,
+        },
+        reply_to,
+        ReplyMsg::IbcTransfer.to_reply_id(),
+    )?;
+    Ok(Response::default().add_submessages(vec![submsg]))
+}
+
+fn register_non_native_rewards_balances_query(
+    deps: DepsMut<NeutronQuery>,
+    info: MessageInfo,
+    denoms: Vec<String>,
+) -> ContractResult<Response<NeutronMsg>> {
+    deps.api.debug(&format!(
+        "WASMDEBUG: register_non_native_rewards_balances_query denoms:{:?}",
+        denoms
+    ));
+    let puppeteer_base = Puppeteer::default();
+    let config = puppeteer_base.config.load(deps.storage)?;
+    cw_ownable::assert_owner(deps.storage, &info.sender)?;
+    let kv_queries = puppeteer_base
+        .kv_queries
+        .range(deps.storage, None, None, Order::Ascending)
+        .collect::<Result<Vec<(u64, KVQueryType)>, _>>()?;
+    let ica = puppeteer_base.ica.get_address(deps.storage)?;
+    let mut messages = vec![];
+    let mut submessages = vec![];
+    for (query_id, query_type) in kv_queries {
+        if query_type == KVQueryType::NonNativeRewardsBalances {
+            messages.push(update_multiple_balances_query_msg(
+                query_id,
+                ica.clone(),
+                denoms.clone(),
+            )?);
+        }
+    }
+    if messages.is_empty() {
+        submessages.push(SubMsg::reply_on_success(
+            new_multiple_balances_query_msg(
```

Comment:
Are you sure this query will work with initia? Maybe it is better to return "not implemented", anyway this code will not work